File: app/src/main/java/com/hjc/reader/ui/gank/child/RecommendFragment.java
package com.hjc.reader.ui.gank.child;

import android.view.View;

import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.hitomi.tilibrary.transfer.TransferConfig;
import com.hitomi.tilibrary.transfer.Transferee;
import com.hjc.baselib.fragment.BaseMvmLazyFragment;
import com.hjc.reader.R;
import com.hjc.reader.bean.response.GankDayBean;
import com.hjc.reader.databinding.FragmentRecommendBinding;
import com.hjc.reader.ui.gank.adapter.RecommendAdapter;
import com.hjc.reader.utils.helper.RouteManager;
import com.hjc.reader.viewmodel.gank.RecommendViewModel;
import com.vansz.glideimageloader.GlideImageLoader;

import java.util.ArrayList;
import java.util.List;

/**
 * @Author: HJC
 * @Date: 2019/1/21 11:29
 * @Description: 每日推荐页面
 */
public class RecommendFragment extends BaseMvmLazyFragment<FragmentRecommendBinding, RecommendViewModel> {

    private RecommendAdapter mAdapter;

    private Transferee transferee;

    public static RecommendFragment newInstance() {
        return new RecommendFragment();
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_recommend;
    }

    @Override
    protected RecommendViewModel getViewModel() {
        return new ViewModelProvider(this).get(RecommendViewModel.class);
    }

    @Override
    protected int getBindingVariable() {
        return 0;
    }

    @Override
    public void initView() {
        setLoadSir(mBindingView.smartRefreshLayout);

        LinearLayoutManager manager = new LinearLayoutManager(mContext);
        mBindingView.rvRecommend.setLayoutManager(manager);

        mAdapter = new RecommendAdapter();
        mBindingView.rvRecommend.setAdapter(mAdapter);

        mAdapter.setAnimationWithDefault(BaseQuickAdapter.AnimationType.SlideInLeft);
    }

    @Override
    public void initData() {
        transferee = Transferee.getDefault(mContext);

        mViewModel.getRecommendData(true);
    }

    @Override
    protected void observeLiveData() {
        mViewModel.getRecommendLiveData().observe(this, data -> {
            mBindingView.smartRefreshLayout.finishRefresh();
            mAdapter.setNewInstance(data);
        });
    }

    @Override
    public void addListeners() {
        mBindingView.smartRefreshLayout.setOnRefreshListener(refreshLayout -> {
            mViewModel.getRecommendData(false);
        });

        mAdapter.setOnItemClickListener((adapter, view, position) -> {
            List<GankDayBean> dataList = mAdapter.getData();
            GankDayBean bean = dataList.get(position);

            int itemViewType = adapter.getItemViewType(position);
            if (itemViewType == GankDayBean.TYPE_IMAGE_ONLY) {
                List<String> imageList = new ArrayList<>();
                imageList.add(bean.getUrl());

                transferee.apply(TransferConfig.build()
                        .setImageLoader(GlideImageLoader.with(getContext()))
                        .setSourceUrlList(imageList)
                        .create()
                ).show();

            } else if (itemViewType == GankDayBean.TYPE_TEXT || itemViewType == GankDayBean.TYPE_IMAGE_ONE || itemViewType == GankDayBean.TYPE_IMAGE_THREE) {
                RouteManager.jumpToWeb(bean.getDesc(), bean.getUrl());
            }
        });
    }

    @Override
    protected void onRetryBtnClick(View v) {
        mViewModel.getRecommendData(true);
    }

    @Override
    public void onSingleClick(View v) {

    }
}

File: module-gank/src/main/java/com/hjc/learn/gank/ui/recommend/RecommendFragment.kt
package com.hjc.learn.gank.ui.recommend

import android.view.View
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.alibaba.android.arouter.facade.annotation.Route
import com.chad.library.adapter.base.BaseQuickAdapter.AnimationType
import com.hitomi.tilibrary.transfer.TransferConfig
import com.hitomi.tilibrary.transfer.Transferee
import com.hjc.learn.gank.R
import com.hjc.learn.gank.adapter.RecommendAdapter
import com.hjc.learn.gank.databinding.GankFragmentRecommendBinding
import com.hjc.learn.gank.viewmodel.RecommendViewModel
import com.hjc.library_base.fragment.BaseLazyFragment
import com.hjc.library_common.router.RouteManager
import com.hjc.library_common.router.path.RouteGankPath
import com.hjc.library_net.model.GankDayBean
import com.vansz.glideimageloader.GlideImageLoader
import java.util.*

/**
 * @Author: HJC
 * @Date: 2019/1/21 11:29
 * @Description: 每日推荐页面
 */
@Route(path = RouteGankPath.URL_FRAGMENT_RECOMMEND)
class RecommendFragment : BaseLazyFragment<GankFragmentRecommendBinding, RecommendViewModel>() {

    private lateinit var mAdapter: RecommendAdapter

    private lateinit var transferee: Transferee

    override fun getLayoutId(): Int {
        return R.layout.gank_fragment_recommend
    }

    override fun createViewModel(): RecommendViewModel {
        return ViewModelProvider(this)[RecommendViewModel::class.java]
    }

    override fun initView() {
        super.initView()

        initLoadSir(mBindingView.smartRefreshLayout)

        val manager = LinearLayoutManager(mContext)
        mBindingView.rvRecommend.layoutManager = manager

        mAdapter = RecommendAdapter()
        mBindingView.rvRecommend.adapter = mAdapter

        mAdapter.setAnimationWithDefault(AnimationType.SlideInLeft)
    }

    override fun initData() {
        transferee = Transferee.getDefault(mContext)

        mViewModel?.getRecommendData(true)
    }

    override fun observeLiveData() {
        super.observeLiveData()

        mViewModel?.run {
            recommendLiveData.observe(this@RecommendFragment) { data ->
                mAdapter.setNewInstance(data)
            }

            refreshData.observe(this@RecommendFragment) { result ->
                if (result) {
                    mBindingView.smartRefreshLayout.finishRefresh()
                }
            }
        }
    }

    override fun addListeners() {
        mBindingView.smartRefreshLayout.setOnRefreshListener { mViewModel?.getRecommendData(false) }

        mAdapter.setOnItemClickListener { _, _, position ->
            val dataList = mAdapter.data
            val bean = dataList[position]

            val itemViewType = mAdapter.getItemViewType(position)
            if (itemViewType == GankDayBean.TYPE_IMAGE_ONLY) {
                val imageList: MutableList<String> = ArrayList()
                bean.url?.let {
                    imageList.add(it)
                }
                transferee.apply(
                    TransferConfig.build()
                        .setImageLoader(GlideImageLoader.with(context))
                        .setSourceUrlList(imageList)
                        .create()
                ).show()
            } else if (itemViewType == GankDayBean.TYPE_TEXT || itemViewType == GankDayBean.TYPE_IMAGE_ONE || itemViewType == GankDayBean.TYPE_IMAGE_THREE) {
                RouteManager.jumpToWeb(bean.desc, bean.url)
            }
        }
    }

    override fun onSingleClick(v: View?) {

    }

    override fun onRetryBtnClick(v: View?) {
        mViewModel?.getRecommendData(true)
    }

    override fun onDestroy() {
        super.onDestroy()
        transferee.destroy()
    }

}